### PR TITLE
[CBRD-21593] Fix keep_from_log_pageid assign

### PR DESCRIPTION
**Pull request by lucianlazar2211**
http://jira.cubrid.org/browse/CBRD-21593

Previously, `keep_from_log_pageid` was set to the `last_blockid + 1`, but it might have happened that the `last_blockid` was refering a page that might reside in a log archive that was previously deleted. Now, in case of empty `vacuum.Data`, `keep_from_log_pageid` is set to `NULL_PAGEID` to ensure no problems are encountered at recovery.

### CODE REVIEW

File: src/query/vacuum.c [hunk 1]
```diff
@@ -5586,9 +5586,17 @@ vacuum_update_keep_from_log_pageid (THREAD_ENTRY * thread_p)
...
   VACUUM_LOG_BLOCKID keep_from_blockid =
     (vacuum_Data.first_page->index_unvacuumed == vacuum_Data.first_page->index_free ?
-     vacuum_Data.last_blockid + 1 : vacuum_Data.first_page->data[vacuum_Data.first_page->index_unvacuumed].blockid);
-  keep_from_blockid = VACUUM_BLOCKID_WITHOUT_FLAGS (keep_from_blockid);
-  vacuum_Data.keep_from_log_pageid = VACUUM_FIRST_LOG_PAGEID_IN_BLOCK (keep_from_blockid);
+     NULL_PAGEID : vacuum_Data.first_page->data[vacuum_Data.first_page->index_unvacuumed].blockid);
+
+  if (keep_from_blockid == NULL_PAGEID)
+    {
+      vacuum_Data.keep_from_log_pageid = NULL_PAGEID;
```
Comment (eseokoh) [inline src/query/vacuum.c]: it would be great if you can also update the above comment :)

File: src/query/vacuum.c [hunk 2]
```diff
@@ -4085,13 +4087,14 @@ vacuum_data_load_and_recover (THREAD_ENTRY * thread_p)
...
-  if (vacuum_Data.last_page->index_unvacuumed == vacuum_Data.last_page->index_free)
+  if (vacuum_is_empty ())
     {
-      /* Empty last page. */
-      assert (vacuum_Data.last_page->index_unvacuumed == 0);
-      /* last_blockid is still saved in the first data entry. */
-      vacuum_Data.last_blockid = vacuum_Data.last_page->data->blockid;
-      assert (vacuum_Data.last_blockid == VACUUM_BLOCKID_WITHOUT_FLAGS (vacuum_Data.last_blockid));
+      /* Get the maximum between what is currently stored in vacuum and the value stored
+       * in the log_Gl header. After a long session in SA_MODE, the vacuum_Data.last_page->data->blockid will not
```
Comment (ribram) [inline src/query/vacuum.c]: Rephrase as "After a long session in SA_MODE, the vacuum_Data.last_page->data->blockid will be outdated. Instead SA_MODE updates log_Gl.hdr.vacuum_last_blockid before removing old archives.".

**Comment by eseokoh** [comment]
I see log compatibility issue on `log_Gl.hdr.vacuum_last_blockid`:
Since we need to also patch 10.1, we have to keep the log image compatibility. 
Fortunately, we may use `reserved_int_1` and `reserved_int_2` for `vacuum_last_blockid`. These reserved fields are now saved as 0. Please think about it.

File: src/query/vacuum.c [hunk 3]
```diff
@@ -665,6 +665,8 @@ static int vacuum_get_first_page_dropped_files (THREAD_ENTRY * thread_p, VPID *
 static bool is_not_vacuumed_and_lost (THREAD_ENTRY * thread_p, MVCC_REC_HEADER * rec_header);
 static void print_not_vacuumed_to_log (OID * oid, OID * class_oid, MVCC_REC_HEADER * rec_header, int btree_node_type);
 
+static bool vacuum_is_empty ();
```
Comment (eseokoh) [inline src/query/vacuum.c]: please fill void for empty parameter.

File: src/query/vacuum.c [hunk 4]
```diff
@@ -7731,3 +7744,21 @@ vacuum_rv_check_at_undo (THREAD_ENTRY * thread_p, PAGE_PTR pgptr, INT16 slotid,
...
   return NO_ERROR;
 }
+
+/* 
+ *	vacuum_is_empty() - Checks if the vacuum is empty.
+ *
+ *	return :- true or false
+ */
+bool
+vacuum_is_empty ()
```
Comment (eseokoh) [inline src/query/vacuum.c]: please fill void for empty parameter.

File: src/query/vacuum.c [hunk 5]
```diff
@@ -5580,15 +5584,24 @@ vacuum_update_keep_from_log_pageid (THREAD_ENTRY * thread_p)
...
+   * the remaining pages, to NULL_PAGEID.
    * If vacuum data is not empty, then we need to preserve the log starting with the first page of first unvacuumed
    * block.
    */
-  VACUUM_LOG_BLOCKID keep_from_blockid =
-    (vacuum_Data.first_page->index_unvacuumed == vacuum_Data.first_page->index_free ?
-     vacuum_Data.last_blockid + 1 : vacuum_Data.first_page->data[vacuum_Data.first_page->index_unvacuumed].blockid);
-  keep_from_blockid = VACUUM_BLOCKID_WITHOUT_FLAGS (keep_from_blockid);
-  vacuum_Data.keep_from_log_pageid = VACUUM_FIRST_LOG_PAGEID_IN_BLOCK (keep_from_blockid);
+  VACUUM_LOG_BLOCKID keep_from_blockid;
```
Comment (eseokoh) [inline src/query/vacuum.c]: a horizontal blank line would help :)

File: src/transaction/log_page_buffer.c [hunk 6]
```diff
@@ -12139,3 +12161,19 @@ logpb_vacuum_reset_log_header_cache (THREAD_ENTRY * thread_p, LOG_HEADER * loghd
...
+ * 
+ * return :- void
+ *
+ * thread_p (in) :- Thread context.
+ * page_id (in)  :- The first page id of the block that must updated to.
+ */
+void
+logpb_update_last_blockid (THREAD_ENTRY * thread_p, LOG_PAGEID page_id)
+{
+  VACUUM_LOG_BLOCKID last_blockid;
```
Comment (eseokoh) [inline src/transaction/log_page_buffer.c]: please also add a blank line.

File: src/transaction/log_impl.h [hunk 7]
```diff
@@ -214,6 +214,7 @@ struct log_header
   INT64 ha_promotion_time;
   INT64 db_restore_time;
   bool mark_will_del;
+  INT64 vacuum_last_blockid;
```
Comment (eseokoh) [inline src/transaction/log_impl.h]: please find my previous comment about compatibility.

File: src/transaction/log_page_buffer.c [hunk 8]
```diff
@@ -6946,6 +6945,19 @@ logpb_remove_archive_logs_exceed_limit (THREAD_ENTRY * thread_p, int max_count)
       if (last_arv_num_to_delete >= first_arv_num_to_delete)
 	{
 	  log_Gl.hdr.last_deleted_arv_num = last_arv_num_to_delete;
+
+	  /* Update the last_blockid needed for vacuum. */
+	  new_page_id = logpb_Arv_page_info_table.page_info[last_arv_num_to_delete + 1].start_pageid;
```
Comment (eseokoh) [inline src/transaction/log_page_buffer.c]: I don't think it is correct. Please notice that `last_arv_num_to_delete` could not be used as index of the page table.

Comment (eseokoh) [reply src/transaction/log_page_buffer.c]: please also notice that you may not find the archive volume from the page table, where overflow of entries of table happens.

File: src/transaction/log_page_buffer.c [hunk 9]
```diff
@@ -6849,6 +6848,7 @@ logpb_remove_archive_logs_exceed_limit (THREAD_ENTRY * thread_p, int max_count)
   int log_max_archives = prm_get_integer_value (PRM_ID_LOG_MAX_ARCHIVES);
   char *catmsg;
   int deleted_count = 0;
+  int local_rear = 0;
```
Comment (eseokoh) [inline src/transaction/log_page_buffer.c]: It is unused :)

File: src/transaction/log_page_buffer.c [hunk 10]
```diff
@@ -6946,6 +6946,19 @@ logpb_remove_archive_logs_exceed_limit (THREAD_ENTRY * thread_p, int max_count)
       if (last_arv_num_to_delete >= first_arv_num_to_delete)
 	{
 	  log_Gl.hdr.last_deleted_arv_num = last_arv_num_to_delete;
+
+	  /* Update the last_blockid needed for vacuum. Get the first page_id of the previously logged archive */
+	  new_page_id = log_Gl.append.prev_lsa.pageid;
+	  if (last_arv_num_to_delete == -1 || new_page_id == 0)
+	    {
+	      /* set as NULL_BLOCKID */
+	      logpb_update_last_blockid (thread_p, NULL_PAGEID);
```
Comment (eseokoh) [inline src/transaction/log_page_buffer.c]: I don't think it is reachable. Can you please confirm?

Comment (lucianlazar2211) [reply src/transaction/log_page_buffer.c]: Indeed, with the current implementation this is no longer reachable. I am removing it.

**Pull request merged**